usr.bin/shuf/shuf.1
." Copyright (c) 2017-2020 Brian Callahan <bcallah@openbsd.org>
."
." Permission to use, copy, modify, and distribute this software for any
." purpose with or without fee is hereby granted, provided that the above
." copyright notice and this permission notice appear in all copies.
."
." THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
." WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
." MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
." ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
." WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
." ACTION OF CONTRACT, NEGLIGENCE OR OTHERWISE) ARISING OUT OF OR IN
." CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
."
.Dd March 20, 2020
.Dt SHUF 1
.Os
.Sh NAME
.Nm shuf
.Nd randomly permute input lines
.Sh SYNOPSIS
.Nm
.Op Fl hv
.Op Fl n Ar count
.Op Fl o Ar outfile
.Op Fl rz
.Op Ar file
.Nm
.Op Fl hv
.Fl e
.Op Fl n Ar count
.Op Fl o Ar outfile
.Op Fl rz
.Op Ar args ...
.Nm
.Op Fl hv
.Fl i Ar lo-hi
.Op Fl n Ar count
.Op Fl o Ar outfile
.Op Fl rz
.Sh DESCRIPTION
.Nm
is a utility that outputs a random permutation of its input lines.
By default,
.Nm
reads from
.Ar stdin
and outputs to
.Ar stdout .
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl e
Use a space-separated list of command-line
.Ar args
as input.
.Pp
Cannot be combined with
.Fl i .
.It Fl h
Print usage information and exit.
.It Fl i Ar lo-hi
Use the range of unsigned integers from
.Ar lo
to
.Ar hi
as input.
.Pp
Cannot be combined with
.Fl e .
.It Fl n Ar count
Print at most
.Ar count
lines.
.It Fl o Ar outfile
Write output to
.Ar outfile
instead of
.Ar stdout .
.It Fl r
Do not permute.
Instead, choose lines at random, with replacement.
When used without
.Fl n ,
.Nm
repeats indefinitely.
.It Fl v
Print version information and exit.
.It Fl z
Delineate lines with the NUL character instead of the newline character.
.El
.Sh EXIT STATUS
The
.Nm
utility exits 0 on success, and >0 if an error occurs.
.Sh EXAMPLES
Flip a coin ten times:
.Pp
.Dl shuf -e -n 10 -r Heads Tails
.Pp
Select five numbers from 10 to 20:
.Pp
.Dl shuf -i 10-20 -n 5
.Pp
Set a random desktop background image from the user's collection:
.Pp
.Dl ls ~/wallpaper | shuf -n 1 | feh --bg-fill
.Pp
Read a shuffled version of this manual page:
.Pp
.Dl shuf shuf.1 | mandoc -mdoc | less
.Sh AUTHORS
.Nm
was written by
.An Brian Callahan Aq Mt bcallah@openbsd.org .
.Sh CAVEATS
This implementation of
.Nm
intentionally omits all long options.
.Pp
The maximum range for
.Ar lo-hi
is
.Dv INT_MAX Ns -1.


NEWS
GNU diffutils NEWS                                    -*- outline -*-

* Noteworthy changes in release 3.3 (2013-03-24) [stable]

** New features

  --new-file (-N) and --unidirectional-new-file now allow comparisons to "-".
  A standard input that's closed acts like a nonexistent file.

  A file name containing spaces, double quotes, backslashes or control
  characters is now encoded in a diff header as a double-quoted C string
  literal.  The escape sequences \\, \", \a, \b, \f, \n, \r, \t, \v and
  \ooo (a three-digit octal number between 0 and 255) are used.

** Packaging

  diffutils is now designed to build with Cygwin or MinGW rather than DJGPP.
  The ms subdirectory has been removed.


* Noteworthy changes in release 3.2 (2011-09-02) [stable]

** Changes in behavior

  --ignore-file-name-case now applies at the top level too.
  For example, "diff dir inIt" might compare "dir/Init" to "inIt".

** New features

  diff and sdiff have a new option --ignore-trailing-space (-Z).

** Packaging

  The texinfo documentation no longer specifies "front-cover" or "back-cover"
  texts, so that it may now be included in Debian's "main" section.


* Noteworthy changes in release 3.1 (2011-08-10) [stable]

** Bug fixes

  diff no longer reports spurious differences merely because two entries
  in the same directory have names that compare equal in the current
  locale, or compare equal because --ignore-file-name-case was given.

* Noteworthy changes in release 3.0 (2010-05-03) [stable]

** Bug fixes

  diff once again prints the required "\ No newline at end of file" line
  when at least one input lacks a newline-at-EOF and the final hunk plus
  context-length aligns exactly with the end of the newline-lacking file.
  [bug introduced between 2.8.7 and 2.9]

** Changes in behavior

  In context-style diffs, diff prints a portion of a preceding "function"
  line for each hunk, with --show-function-line=RE (-F) or
  --show-c-function (-p).  Now, it trims leading blanks from such lines
  before extracting a prefix.  This is useful especially when a function
  line is so far indented that the name itself would be truncated or not
  included in the limited-width substring that diff appends.

  diff once again reports a difference with the diagnostic
  "Binary files A and B differ" when at least one of the files
  appears to be binary.  From 2.8.4 through diffutils-2.9, it printed
  "Files A and B differ".


* Noteworthy changes in release 2.9 (2010-02-11) [stable]

** New features

  New diff option --suppress-blank-empty.

  Bring back support for 'diff -NUM', where NUM is a number,
  even when conforming to POSIX 1003.1-2001.  This change reverts to
  the behavior of GNU diff 2.7 and earlier.  This is a change only
  when conforming to POSIX 1003.1-2001; there is no effect when
  conforming to older POSIX versions.

  This change is in response to decisions taken in the January 2005
  Austin Group standardization meeting.  For more details, please see
  "Utility Syntax Guidelines" in the Minutes of the January 2005
  Meeting <http://www.opengroup.org/austin/docs/austin_239.html>.

  sdiff now understands '1' and '2' as synonyms for 'l' and 'r'.

** Changes in behavior

  sdiff and diff3 now invoke diff, not $(bindir)/diff

** Administrivia

  New discussion and bug-reporting address: bug-diffutils@gnu.org

  updated gnulib support


* Noteworthy changes in release 2.8.7 (2004-04-13) [stable]

  Version 2.8.7 contains no user-visible changes.

User-visible changes in version 2.8.6:

* New diff3 option --strip-trailing-cr.

* With -N and -P, inaccessible empty regular files (the kind of files
  that 'patch' creates to indicate nonexistent backups) are now
  treated as nonexistent when they are in the 'backup' file position.

* If multiple SKIP values are given to cmp, e.g., 'cmp -i 10 -i 20',
  cmp now uses the maximal value instead of the last one.

* diff now omits the ".000000000" on hosts that do not support
  fractional time stamps.

Version 2.8.5 was not publicly released.

User-visible changes in version 2.8.4:

* Diff now simply prints "Files A and B differ" instead of "Binary
  files A and B differ".  The message is output if either A or B
  appears to be a binary file, and the old wording was misleading
  because it implied that both files are binary, which is not
  necessarily the case.

User-visible changes in version 2.8.3:

* New locale: en_US.

User-visible changes in version 2.8.2:

* New diff and sdiff option:
  --tabsize=COLUMNS
* If --ignore-space-change or --ignore-all-space is also specified,
  --ignore-blank-lines now considers lines to be empty if they contain
  only white space.
* More platforms now handle multibyte characters correctly when
  excluding files by name (diff -x and -X).
* New locales: hu, pt_BR.

User-visible changes in version 2.8.1:

* Documentation fixes.

User-visible changes in version 2.8:

* cmp and diff now conform to POSIX 1003.1-2001 (IEEE Std 1003.1-2001)
  if the underlying system conforms to POSIX and if the _POSIX2_VERSION
  environment variable is set to 200112.  Conformance removes support
  for 'diff -NUM', where NUM is a number.  Use -C NUM or -U NUM instead.
* cmp now supports trailing operands SKIP1 and SKIP2, like BSD cmp.
* cmp -i or --ignore-initial now accepts SKIP1:SKIP2 option value.
* New cmp option: -n or --bytes.
* cmp's old -c or --print-chars option has been renamed;
  use -b or --print-bytes instead.
* cmp now outputs "byte" rather than "char" outside the POSIX locale.
* cmp -l's index column width now adjusts to fit larger (or smaller) files.
* cmp -l -s and cmp -s -l are not allowed.  Use cmp -s or cmp -l instead.
* diff uses ISO 8601 style time stamps for output times (e.g. "2001-11-23
  16:44:36.875702460 -0800") unless in the C or POSIX locale and the
  -c style is specified.
* diff's -I and -F options use the regexp syntax of grep, not of Emacs.
* diff now accepts multiple context arguments, and uses their maximum value.

usr.bin/shuf/shuf.c
/*
 * Copyright (c) 2017-2020 Brian Callahan <bcallah@openbsd.org>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHERWISE) ARISING OUT OF OR IN
 * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <sys/capsicum.h>

#include <err.h>
#include <errno.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <capsicum_helpers.h>

static FILE    *ofile;
static char 	delimiter = '\n';
static int 	eflag, most = -1, rflag, zflag;

static void __dead2 usage(void);
static void __dead2 version(void);
static void enter_capability_mode(FILE *);
static void limit_standard_fd(int, int, int);
static void limit_rights(int, const cap_rights_t *, const char *, int);
static void printshuf(char *args[], int ind);
static void randomshuf(int argn, char *args[]);
static void shuf(int argn, char *args[]);
static void shuffile(const char *input, int argn, size_t inputlen);
static void shufintegers(int range, int lo);

/*
 * shuf -- output a random permutation of input lines
 */

static void
limit_rights(int fd, const cap_rights_t *rights, const char *name,
    int ignore_ebadf)
{

	if (caph_rights_limit(fd, rights) < 0) {
		if (ignore_ebadf && errno == EBADF)
			return;
		err(1, "unable to limit rights for %s", name);
	}
}

static void
limit_standard_fd(int fd, int ifd, int ofd)
{
	cap_rights_t rights;
	int needs_fstat;

	cap_rights_init(&rights);

	needs_fstat = 0;
	if (fd == ifd) {
		cap_rights_set(&rights, CAP_READ);
		needs_fstat = 1;
	}
	if (fd == ofd || fd == STDERR_FILENO) {
		cap_rights_set(&rights, CAP_WRITE);
		needs_fstat = 1;
	}
	if (needs_fstat)
		cap_rights_set(&rights, CAP_FSTAT);

	limit_rights(fd, &rights, "standard descriptor", 1);
}

static void
enter_capability_mode(FILE *ifile)
{
	cap_rights_t rights_ro, rights_wo;
	int ifd, ofd;

	ifd = ifile == NULL ? -1 : fileno(ifile);
	ofd = fileno(ofile);

	cap_rights_init(&rights_ro, CAP_FSTAT, CAP_READ);
	cap_rights_init(&rights_wo, CAP_FSTAT, CAP_WRITE);

	limit_standard_fd(STDIN_FILENO, ifd, ofd);
	limit_standard_fd(STDOUT_FILENO, ifd, ofd);
	limit_standard_fd(STDERR_FILENO, ifd, ofd);

	if (ifd > STDERR_FILENO)
		limit_rights(ifd, &rights_ro, "input", 0);
	if (ofd > STDERR_FILENO)
		limit_rights(ofd, &rights_wo, "output", 0);

	caph_cache_catpages();
	if (caph_enter() < 0)
		err(1, "unable to enter capability mode");
}

static void
printshuf(char *args[], int ind)
{
	size_t size;

	for (size = 0; args[ind][size] != delimiter; size++)
		;

	fwrite(args[ind], eflag ? size : size + 1, 1, ofile);

	if (eflag) {
		if (zflag)
			fputc('\0', ofile);
		else
			fputc('\n', ofile);
	}
}

static void
randomshuf(int argn, char *args[])
{

	while ((most == -1 ? 1 : most-- > 0))
		printshuf(args, arc4random_uniform(argn));
}

static void
shuf(int argn, char *args[])
{
	char   *argt;
	int 	i, j;

	for (i = argn - 1; i > 0; i--) {
		j = arc4random_uniform(i + 1);
		argt = args[j];
		args[j] = args[i];
		args[i] = argt;
	}

	most = (most == -1 ? argn : most > argn ? argn : most);

	for (i = 0; i < most; i++)
		printshuf(args, i);
}

static void
shuffile(const char *input, int argn, size_t inputlen)
{
	const char     *s;
	char	      **args = NULL, *argt;
	int 			i = 0;
	size_t 		len;

	if ((args = reallocarray(args, argn, sizeof(char *))) == NULL)
		err(1, "could not create shuffile buffer");

	while (i < argn) {
		for (s = input; *s != delimiter; s++) {
			if (inputlen-- == 0)
				break;
		}
		len = s - input;

		if ((args[i] = malloc(len + 1)) == NULL)
			err(1, "could not create shuffile item buffer");
		argt = args[i++];

		while (len-- > 0)
			*argt++ = *input++;
		*argt++ = delimiter;

		input++;
		--inputlen;
	}

	most = (most < argn ? most : argn);

	if (rflag)
		randomshuf(argn, args);
	else
		shuf(argn, args);

	for (i = 0; i < argn; i++) {
		free(args[i]);
		args[i] = NULL;
	}

	free(args);
	args = NULL;
}

static void
shufintegers(int range, int lo)
{
	int    *args = NULL, argt, i, j;

	if ((args = reallocarray(args, range, sizeof(int))) == NULL)
		err(1, "range size will exhaust memory");

	if (rflag) {
		while ((most == -1 ? 1 : most-- > 0)) {
			fprintf(ofile, "%u%c",
				arc4random_uniform(range) + lo, delimiter);
		}
	} else {
		for (i = 0; i < range; i++)
			args[i] = lo + i;

		for (i = range - 1; i > 0; i--) {
			j = arc4random_uniform(i + 1);
			argt = args[j];
			args[j] = args[i];
			args[i] = argt;
		}

		most = (most == -1 ? range : most > range ? range : most);

		for (i = 0; i < most; i++)
			fprintf(ofile, "%d%c", args[i], delimiter);
	}

	free(args);
	args = NULL;
}

static void __dead2
usage(void)
{
	const char     *name;

	name = getprogname();

	fprintf(stderr, "usage: %s [-hv] [-n count] [-o file] [-rz] [file]\n"
		"       %s [-hv] -e [-n count] [-o file] [-rz] [args ...]\n"
		"       %s [-hv] -i lo-hi [-n count] [-o file] [-rz]\n",
		name, name, name);

	exit(1);
}

static void __dead2
version(void)
{

	fputs("shuf 3.0\n"
	      "Copyright (c) 2017-2020 Brian Callahan <bcallah@openbsd.org>\n"
	      "\nPermission to use, copy, modify, and distribute this software"
	      " for any\npurpose with or without fee is hereby granted, "
	      "provided that the above\ncopyright notice and this permission "
	      "notice appear in all copies.\n\nTHE SOFTWARE IS PROVIDED "
	      " AS\nIS\" AND THE AUTHOR DISCLAIMS ALL WARRANTIES"
	      " WITH REGARD TO\nTHIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF"
	      "\nMERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE "
	      "LIABLE FOR\nANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL "
	      "DAMAGES OR ANY DAMAGES\nWHATSOEVER RESULTING FROM LOSS OF USE, "
	      "DATA OR PROFITS, WHETHER IN AN\nACTION OF CONTRACT, NEGLIGENCE "
	      "OR OTHER TORTIOUS ACTION, ARISING OUT OF\nOR IN CONNECTION "
	      "WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.\n", stderr);

	exit(1);
}

int
main(int argc, char *argv[])
{
	FILE	       *ifile;
	const char     *errstr;
	char	       *argp, *buf, *nbuf;
	int 			argn = 1, ch, hi = 0, lo = 0, prev = 1;
	int 			iflag = 0, oflag = 0;
	size_t 		buflen = 0, bufsize = 8192, nbufsize;

	while ((ch = getopt(argc, argv, "ehi:n:o:rvz")) != -1) {
		switch (ch) {
		case 'e':
			eflag = 1;
			break;
		case 'h':
			usage();
		case 'i':
			if (iflag++)
				errx(1, "cannot have multiple -i options");

			if ((argp = strchr(optarg, '-')) == NULL)
				errx(1, "must provide lo and hi for -i option");
			*argp = '\0';

			lo = strtonum(optarg, 0, INT_MAX, &errstr);
			if (errstr != NULL)
				errx(1, "invalid or out of range lo");

			hi = strtonum(argp + 1, 0, INT_MAX, &errstr);
			if (errstr != NULL)
				errx(1, "invalid or out of range hi");

			if (lo >= hi)
				errx(1, "lo is greater than or equal to hi");
			if (hi == INT_MAX && lo == 0)
				errx(1, "lo-hi range too large for -i option");
			break;
		case 'n':
			most = strtonum(optarg, 0, INT_MAX, &errstr);
			if (errstr != NULL) {
				errx(1,
				     "-n count must be from 0 to %d, not %s",
				     INT_MAX, optarg);
			}
			break;
		case 'o':
			if (oflag++)
				errx(1, "cannot have multiple -o options");

			if ((ofile = fopen(optarg, "w")) == NULL)
				err(1, "could not open output file %s", optarg);

			break;
		case 'r':
			rflag = 1;
			break;
		case 'v':
			version();
		case 'z':
			zflag = 1;
			delimiter = '\0';
			break;
		default:
			usage();
		}
	}
	argc -= optind;
	argv += optind;

	if (eflag && iflag)
		errx(1, "cannot combine -e and -i options");

	if (oflag == 0)
		ofile = stdout;

	if (eflag == 0 && ((iflag == 1 && argc != 0) || argc > 1))
		errx(1, "extra operand '%s'", iflag == 1 ? *argv : *++argv);

	if (eflag) {
		enter_capability_mode(NULL);
		if (argc == 0)
			goto out;
		delimiter = '\0';
		if (rflag)
			randomshuf(argc, argv);
		else
			shuf(argc, argv);
		goto out;
	}

	if (iflag) {
		enter_capability_mode(NULL);
		shufintegers(hi - lo + 1, lo);
		goto out;
	}

	/* Open input file before entering capability mode */
	if (argc == 0 || (argv[0][0] == '-' && argv[0][1] == '\0')) {
		ifile = stdin;
	} else {
		if ((ifile = fopen(*argv, "r")) == NULL)
			err(1, "could not open %s", *argv);
	}

	enter_capability_mode(ifile);

	if ((buf = malloc(bufsize)) == NULL)
		err(1, "could not create initial shuffle buffer");

	while ((ch = fgetc(ifile)) != EOF) {
		buf[buflen++] = ch;

		if (buflen == bufsize) {
			nbufsize = bufsize << 1;
			if ((nbuf = realloc(buf, nbufsize)) == NULL) {
				free(buf);
				buf = NULL;
				err(1, "could not resize shuffle buffer");
			}
			buf = nbuf;
			bufsize = nbufsize;
		}
		if (prev == delimiter)
			++argn;
		prev = ch;
	}
	buf[buflen] = '\0';

	if (ifile != stdin)
		fclose(ifile);

	if (buflen > 0)
		shuffile(buf, argn, buflen);

	free(buf);
	buf = NULL;

out:
	if (oflag)
		fclose(ofile);

	return 0;
}
